### PR TITLE
feat: Remove binaries from argoexec image. Fixes #7486

### LINKED ISSUE #7253: Distribute more secure and lightweight docker images with docker-slim
# Summary

What change needs making?

Switch to use docker-slim for image building. https://github.com/docker-slim/docker-slim

# Use Cases

When would you use this?

I liked the security part of it in particular. Here's what it claims:
>  docker-slim will optimize and secure your containers by understanding your application and what it needs using various analysis techniques. It will throw away what you don't need, reducing the attack surface of your container.

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### LINKED ISSUE #7486: Make argoexec image scratch/distroless to improve security
# Summary

Argoexec image has lots of components, which makes it vulnerable to various security findings. With emissary executor though it's possible to make the image distroless.

# Use Cases

Workflows using emissary executor.

- [x] #7795
- [x] #7796
- [x] #7797
- [x] #7798
- [x] #7799
- [x] #7800

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### LINKED ISSUE #7796: Replace use of `curl` binary in the HTTP artifact driver with pure Go

### LINKED ISSUE #7797: Replace use of `kubectl` with pure Go in the resource template code

### LINKED ISSUE #7798: Replace use of `git` binary in Git artifact driver
(does Golang Git client support submodules)

### LINKED ISSUE #7799: Replace the `jq` with pure Go

### LINKED ISSUE #7800: Replace use of the `tar` with pure Go in the Kubernetes API executor
This maybe obviated by #7802 .

### LINKED ISSUE #8291: Remove `docker` binary from `argoexec` (i.e. Dockerfile)
This got overlooked in the code review for #8212. This binary can be removed, for security.

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>

Fixes #8291 
Fixes #7800
Fixes #7798
Fixes #7796
Fixes #7486

`argoexec` still runs as root with read-write FS, but is much lower risk.

This PR obviates the following:

Closes #7253
Closes #7799 
Closes #7983 
Closes #7797 

Why? The big security issue with both alpine and debian base images is that, when compromised by an attacker:

* As root, can install any app you want. 
* Anything installed can be run.

Using distroless/scratch only things pre-installed can be run: `argoexec`, `jq` and `kubectl`. Naturally, `kubectl` is still an bad escalation if the pod is over-permissioned. 

Next steps:

- #6895

### CODE REVIEW

**Comment by terrytangyuan** [comment]
Thanks let me know when ready for review.

**Comment by alexec** [comment]
This PR should have a tough review.

File: workflow/artifacts/git/git.go [hunk 1]
```diff
@@ -45,38 +44,29 @@ func GetUser(url string) string {
...
 		err = ioutil.WriteFile(privateKeyFile.Name(), []byte(g.SSHPrivateKey), 0o600)
 		if err != nil {
-			return nil, nil, nil, err
+			return nil, nil, err
 		}
 		auth := &ssh2.PublicKeys{User: sshUser, Signer: signer}
 		if g.InsecureIgnoreHostKey {
 			auth.HostKeyCallback = ssh.InsecureIgnoreHostKey()
 		}
-		args := []string{"ssh", "-i", privateKeyFile.Name()}
```
Comment (alexec) [inline workflow/artifacts/git/git.go:66]: this is always been broken

File: workflow/artifacts/git/git.go [hunk 2]
```diff
@@ -173,50 +158,38 @@ func (g *ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string) error {
...
-			submodulesCmd := exec.Command("git", "submodule", "update", "--init", "--recursive", "--force")
-			submodulesCmd.Dir = path
-			submodulesCmd.Env = env
-			submoduleOutput, err := submodulesCmd.Output()
+			s, err := w.Submodules()
 			if err != nil {
-				return g.error(err, cmd)
+				return err
+			}
+			if err := s.Update(&git.SubmoduleUpdateOptions{
```
Comment (alexec) [inline workflow/artifacts/git/git.go:177]: needs tests

File: workflow/artifacts/git/git.go [hunk 3]
```diff
@@ -173,50 +158,38 @@ func (g *ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string) error {
...
 	if inputArtifact.Git.Revision != "" {
-		// We still rely on forking git for checkout, since go-git does not have a reliable
-		// way of resolving revisions (e.g. mybranch, HEAD^, v1.2.3)
-		rev := getRevisionForCheckout(inputArtifact.Git.Revision)
-		log.Info("Checking out revision ", rev)
-		cmd := exec.Command("git", "checkout", rev)
-		cmd.Dir = path
-		cmd.Env = env
-		output, err := cmd.Output()
+		h, err := repo.ResolveRevision(plumbing.Revision(inputArtifact.Git.Revision))
```
Comment (alexec) [inline workflow/artifacts/git/git.go:161]: needs tests: mybranch, HEAD^, v1.2.3

File: workflow/artifacts/git/git.go [hunk 4]
```diff
@@ -112,111 +77,87 @@ func (g *ArtifactDriver) Save(string, *wfv1.Artifact) error {
...
+		if err = r.Fetch(opts); isFetchErr(err) {
+			return fmt.Errorf("failed to fetch %v: %w", refSpecs, err)
+		}
+	}
+	w, err := r.Worktree()
+	if err != nil {
+		return fmt.Errorf("failed to get work tree: %w", err)
+	}
+	if a.Revision != "" {
+		if err := r.Fetch(&git.FetchOptions{RefSpecs: []config.RefSpec{"refs/heads/*:refs/heads/*"}}); isFetchErr(err) {
```
Comment (terrytangyuan) [inline workflow/artifacts/git/git.go:130]: Should this be `refs/heads/*:refs/remotes/origin/*`?

File: workflow/artifacts/git/git_test.go [hunk 5]
```diff
@@ -1,105 +1,128 @@
...
+			assert.FileExists(t, path+"/README.md")
 		})
-	}
+		t.Run("HTTPS", func(t *testing.T) {
+			token := os.Getenv("PERSONAL_ACCESS_TOKEN")
+			if token == "" {
+				t.SkipNow()
+			}
+			driver := &ArtifactDriver{Username: "alexec", Password: token}
+			assert.NoError(t, load(driver, &wfv1.GitArtifact{Repo: "https://github.com/argoproj-labs/private-test-repo.git"}))
```
Comment (terrytangyuan) [inline workflow/artifacts/git/git_test.go:43]: This test requires the existence of this private repo and requires a personal access token. Is there a better way?

File: workflow/artifacts/git/git.go (hunk 4 above)
Comment (alexec) [reply workflow/artifacts/git/git.go:130]: Hmm. Passes the tests. Maybe?

File: workflow/artifacts/git/git_test.go [hunk 6]
```diff
@@ -1,105 +1,128 @@
...
+		})
+		t.Run("LocalBranch", func(t *testing.T) {
+			assert.NoError(t, load(driver, &wfv1.GitArtifact{Repo: "https://github.com/argoproj-labs/test-repo.git", Revision: "my-branch"}))
+			assert.FileExists(t, path+"/my-branch")
+		})
+		t.Run("Tag", func(t *testing.T) {
+			assert.NoError(t, load(driver, &wfv1.GitArtifact{Repo: "https://github.com/argoproj-labs/test-repo.git", Revision: "v0.0.0"}))
+			assert.FileExists(t, path+"/README.md")
+		})
+	})
```
[the remark this answers is not in the document]
Comment (alexec) [reply workflow/artifacts/git/git_test.go:107]: Good idea

File: workflow/artifacts/git/git_test.go (hunk 5 above)
Comment (alexec) [reply workflow/artifacts/git/git_test.go:43]: this only runs locally

File: workflow/artifacts/git/git_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexec) [reply workflow/artifacts/git/git_test.go:107]: I've locked the repos.

**Comment by alexec** [comment]
changes made - ready for review again

**Review by terrytangyuan — APPROVED** [review]
LGTM.

**Comment by sigwinch28** [comment]
The removal of the `git` binary means that `file://` scheme git artifacts are no longer supportedbecause `go-git` is not _pure_ go: it relies on the `git` binary for the `file://` cheme: https://github.com/go-git/go-git/blob/master/COMPATIBILITY.md#transport-schemes

> Warning: this is not pure Golang. This shells out to the git binary.

**Pull request merged**